src/graph/graph.cc

```diff
@@ -20,8 +21,10 @@ Graph::Graph(IdArray src_ids, IdArray dst_ids, IdArray edge_ids, size_t num_node
   CHECK(IsValidIdArray(edge_ids));
   this->AddVertices(num_nodes);
   num_edges_ = src_ids->shape[0];
-  CHECK(static_cast<int64_t>(num_edges_) == dst_ids->shape[0]) << "vectors in COO must have the same length";
-  CHECK(static_cast<int64_t>(num_edges_) == edge_ids->shape[0]) << "vectors in COO must have the same length";
+  CHECK(static_cast<int64_t>(num_edges_) == dst_ids->shape[0])
```

Comment:
Later on I shall change `NDArray` to use `uint64_t` shapes.  These `static_cast`'s are hurting my eyes.

Reply:
Sparse Attention?

Reply:
I guess a better way is to always use `int64_t`. We only use `uint64_t` when we are very sure about it.

Reply:
> I guess a better way is to always use `int64_t`. We only use `uint64_t` when we are very sure about it.

We are very sure that *pretty much everywhere* should have used `uint64_t` (like `num_edges_`, `NumVertices()`, and a bunch of `vector.size()`'s).  The only issue is that `DLTensor`'s have *signed* integer shapes, which requires us to do `static_cast`'s or punches your face with compiler warnings otherwise.

I don't know why they made this decision; what does a negative shape stand for?

Reply:
Yeah, I think so. We could do that later.